soccer/gameplay/situational_play_selection.py

```diff
@@ -298,60 +318,67 @@ def isMidfield(self):
-    # Preemption is still an open question but this is a prototype of
-    # of how a non-invasive preemption system might work
-    #
-    def updatePreempt(self):
-        if (self.lastSituation != self.currentSituaion and
-                self.situationChangeTime == None):
-            self.situationChangeTime = time.time()
+    ##Returns true if we are in any of the situations in the passed list
+    #passed situation list must be enums
+    def isSituations(self, situations):
+        #This could be pretty easily inverted into a contains call
+        for g in situations:
+            if (self.isSituation(g)):
+                return True
 
-        if (self.currentPreempt):
-            self.currentPreempt = False
-        elif (self.situationChangeTime != None):
-            if (abs(time.time() - self.situationChangeTime) >
-                    self.playPreemptTime):
-                self.situationChangeTime = None
-                self.currentPreempt = True
-                self.LastSituation = self.currentSituaion
+        return False
 
-    '''def addPreempt(play) possibly a function to add transition out of every state to the completed state with preemptPlay as the lambda
-        for g in states:
-            play.add_transition(g -> completed , preemptPlay)'''
+    ##
+    # 
+    # Will determine if the current play has lasted too long over a situation change
+    #   and needs to be preempted, it will call try_preempt
+    def updatePreempt(self):
+
+        #Get the current play, may want to add a getter, as this is a "private" variable
+        currentPlay = main._root_play.play
 
-    ##A function to determine if the currently running play should be preempted
-    def preemptPlay(self):
-        return self.currentPreempt
+        #If the last situation is not the current situation, and the situationChanged flag
+        #is not already up, put it up and record the time, and change last situation to be the current situation
+        if (self.lastSituation != self.currentSituation and
+                not self.situationChanged):
+            self.situationChangeTime = time.time()
+            self.lastSituation = self.currentSituation
+            self.situationChanged = True
+
+        #If the play changes, put the situationChanged flag down and change lastPlay
+        if (self.lastPlay != currentPlay):
+            self.lastPlay = currentPlay
+            self.situationChanged = False
+
+        #If the situation has changed and its been longer than situationChange
+        #time without a play change, try to preempt the current play
+        if (self.situationChanged and
+                abs(time.time() - self.situationChangeTime) >
+                self.preemptTime):
+            #Check to see if the play is a standard play before trying to preempt
+            if (currentPlay != None and isinstance(
+                    currentPlay, standard_play.StandardPlay) and self.enabled):
+                currentPlay.try_preempt()
```

Comment:
Shouldn't this use the result from `try_preempt()`? Or does it just use the fact that the play has now been terminated if preemption succeeded?

Reply:
Current behavior is just that it will continuously call try_preempt() until the play changes without regard to the return value of try_preempt(). Which I think is fine for the time being.

Reply:
Okay. So the play is responsible for deciding if it's terminated, and then situational play selection chooses a new one if it terminates itself.

Sounds good to me.